src/stan/mcmc/hmc/nuts/base_parallel_nuts.hpp

```diff
@@ -0,0 +1,780 @@
+    double sum_metro_prob = 0;
+
+    std::vector<bool> fwd_direction(this->max_depth_);
+
+    for (std::size_t i = 0; i != this->max_depth_; ++i)
+      fwd_direction[i] = this->rand_uniform_() > 0.5;
+
+    const std::size_t num_fwd
+        = std::accumulate(fwd_direction.begin(), fwd_direction.end(), 0);
+    const std::size_t num_bck = this->max_depth_ - num_fwd;
+
+    /*
+    std::cout << "sampled turns: ";
+    for (std::size_t i = 0; i != this->max_depth_; ++i) {
+      if(fwd_direction[i])
+        std::cout <<  "+,";
+      else
+        std::cout << "-,";
+    }
+    std::cout << std::endl;
+    */
+
+    tbb::concurrent_vector<extend_tree_t> ends(
+        this->max_depth_, std::make_tuple(true, 0, Eigen::VectorXd(),
+                                          Eigen::VectorXd(), z_sample, 0, 0.0));
+    tbb::concurrent_vector<bool> valid_subtree_fwd(num_fwd, true);
+    tbb::concurrent_vector<bool> valid_subtree_bck(num_bck, true);
+
+    // HACK!!!
+    /*
+    callbacks::logger logger_fwd;
+    callbacks::logger logger_bck;
+    */
+    // build TBB flow graph
+    tbb::flow::graph g;
+
+    // add nodes which advance the left/right tree
+    using tree_builder_t = tbb::flow::continue_node<tbb::flow::continue_msg>;
+
+    tbb::concurrent_vector<std::size_t> all_builder_idx(this->max_depth_);
+    tbb::concurrent_vector<tree_builder_t> fwd_builder;
+    fwd_builder.reserve(this->max_depth_);
+    tbb::concurrent_vector<tree_builder_t> bck_builder;
+    bck_builder.reserve(this->max_depth_);
+    using builder_iter_t = tbb::concurrent_vector<tree_builder_t>::iterator;
+
+    // now wire up the fwd and bck build of the trees which
+    // depends on single-core or multi-core run
+
+    // TODO: the extenders should also check for a global flag if
+    // we want to keep running
+    // TODO: We should also just run depth = 0 outside the loop to avoid the
+    // if statement here
+    for (std::size_t depth = 0, fwd_idx = 0, bck_idx = 0;
+         depth != this->max_depth_; ++depth) {
+      if (fwd_direction[depth]) {
+        builder_iter_t fwd_iter = fwd_builder.emplace_back(
+            g, [&, depth, fwd_idx](tbb::flow::continue_msg) {
+              // std::cout << "fwd turn at depth " << depth;
+              bool valid_parent
+                  = fwd_idx == 0 ? true : valid_subtree_fwd[fwd_idx - 1];
+              if (valid_parent) {
+                // std::cout << " yes, here we go!" << std::endl;
+                ends[depth] = extend_tree(depth, tree_fwd, z_fwd, logger);
+                valid_subtree_fwd[fwd_idx] = std::get<0>(ends[depth]);
+              } else {
+                valid_subtree_fwd[fwd_idx] = false;
+              }
+              // std::cout << " nothing to do." << std::endl;
+            });
+        if (fwd_idx != 0) {
+          // in this case this is not the starting node, we
+          // connect this with its predecessor
+          tbb::flow::make_edge(*(fwd_iter - 1), *fwd_iter);
```

Comment:
Is this dependency on past forward iterations because we want to check if `valid_subtree_fwd[fwd_idx - 1]` is false and if so do no work? Kind of a wacky idea, but what if we made `valid_parent` a reference to a bool and passed it to extend_tree so we keep checking it at each successive call to `extend_tree`/`build_tree`? Then when a parent fails the children can stop as well

EDIT: I'll leave this comment in case I'm wrong but staring at `build_tree` idt its safe to do what I said above since z_fwd is modified at each depth